src/Nethermind/Nethermind.Merge.AuRa/Shutter/ShutterTxSource.cs

```diff
@@ -0,0 +1,335 @@
+        _abiEncoder = abiEncoder;
+        _auraConfig = auraConfig;
+        _specProvider = specProvider;
+        _logger = logManager.GetClassLogger();
+        _ethereumEcdsa = ethereumEcdsa;
+        _eon = eon;
+        _sequencerContract = new(auraConfig.ShutterSequencerContractAddress, logFinder, filterStore);
+        _validatorRegistryContractAddress = new(_auraConfig.ShutterValidatorRegistryContractAddress);
+        _validatorsInfo = validatorsInfo;
+        _encryptedGasLimit = _auraConfig.ShutterEncryptedGasLimit;
+        _instanceId = _auraConfig.ShutterInstanceID;
+    }
+
+    public IEnumerable<Transaction> GetTransactions(BlockHeader parent, long gasLimit, PayloadAttributes? payloadAttributes = null)
+    {
+        // assume validator will stay registered
+        if (!_validatorsRegistered)
+        {
+            if (!IsRegistered(parent))
+            {
+                return [];
+            }
+
+            _validatorsRegistered = true;
+        }
+
+        ulong nextSlot = GetNextSlot();
+        if (_loadedTransactionsSlot == nextSlot) return _loadedTransactions;
+        if (_logger.IsWarn) _logger.Warn($"Decryption keys not received for slot {nextSlot}, cannot include Shutter transactions.");
+        if (_logger.IsDebug) _logger.Debug($"Current Shutter decryption keys stored for slot {_loadedTransactionsSlot}");
+        return [];
+    }
+
+    public void OnDecryptionKeysReceived(Dto.DecryptionKeys decryptionKeys)
+    {
+        if (decryptionKeys.Gnosis.Slot <= _loadedTransactionsSlot)
+        {
+            if (_logger.IsDebug) _logger.Debug($"Skipping Shutter decryption keys from slot {decryptionKeys.Gnosis.Slot}, keys currently stored for slot {_loadedTransactionsSlot}.");
+            return;
+        }
+
+        ShutterEon.Info? eonInfo = _eon.GetCurrentEonInfo();
+        if (eonInfo is null)
+        {
+            if (_logger.IsDebug) _logger.Debug("Cannot check Shutter decryption keys, eon info was not found.");
+            return;
+        }
+
+        if (_logger.IsDebug) _logger.Debug($"Checking Shutter decryption keys instanceID: {decryptionKeys.InstanceID} eon: {decryptionKeys.Eon} #keys: {decryptionKeys.Keys.Count} #sig: {decryptionKeys.Gnosis.Signatures.Count()} #txpointer: {decryptionKeys.Gnosis.TxPointer} #slot: {decryptionKeys.Gnosis.Slot}");
+
+        if (CheckDecryptionKeys(decryptionKeys, eonInfo.Value))
+        {
+            if (_logger.IsInfo) _logger.Info($"Validated Shutter decryption key for slot {decryptionKeys.Gnosis.Slot}");
+
+            List<SequencedTransaction> sequencedTransactions = GetNextTransactions(decryptionKeys.Eon, decryptionKeys.Gnosis.TxPointer);
+            if (_logger.IsInfo) _logger.Info($"Got {sequencedTransactions.Count} transactions from Shutter mempool...");
+
+            _loadedTransactions = DecryptSequencedTransactions(sequencedTransactions, decryptionKeys);
+            _loadedTransactionsSlot = decryptionKeys.Gnosis.Slot;
```

Comment:
This is a similar race condition to what you had in `ShutterEon`. It is possible for one thread to read tx with mismatched slot.